### PR TITLE
Feat: Streaming answer

### PR DESCRIPTION
**Pull request by micache**
## 🎯 Description

_Describe the big picture of your changes here to communicate to the maintainers why we should accept this pull request._

_If it fixes a bug or resolves a feature request, be sure to link to that issue._

## ☑️ Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR._
- [x] I have added Labels for this PR
- [x] Lint and unit tests pass locally with my changes
- [ ] I have added tests that prove my fix is effective, or my feature works
- [ ] I have added necessary documentation (if appropriate)
- [x] Any dependent changes have been merged and published in downstream modules

## ✍ Further comments

_If this is a relatively large or complex change, kick off the discussion by explaining why you chose the solution you did and what alternatives you considered, etc..._

### CODE REVIEW

File: app/llama_sensei/backend/qa/generate_answer.py [hunk 1]
```diff
@@ -1,31 +1,46 @@
...
     DocumentProcessor,
 )
 from ragas import evaluate
 from ragas.metrics import faithfulness
+from sklearn.metrics.pairwise import cosine_similarity
 
 MODEL = "llama3-70b-8192"
+# device = "cuda" if torch.cuda.is_available() else "cpu"
+# EMBEDDING_LLM = SentenceTransformer("all-MiniLM-L12-v2", trust_remote_code=True).to(device)
+EMBEDDING_LLM = SentenceTransformerEmbeddings(
```
Comment (Huy1711) [inline app/llama_sensei/backend/qa/generate_answer.py]: Please change to EMBEDDING_LLM = "all-MiniLM-L12-v2" and initialize the embedder below (inside the GenerateRAGAnswer __init__ function)

File: app/llama_sensei/backend/qa/generate_answer.py [hunk 2]
```diff
@@ -1,31 +1,46 @@
...
+    model_name="all-MiniLM-L12-v2", model_kwargs={"trust_remote_code": True}
+)
 
 
 class GenerateRAGAnswer:
     def __init__(self, query: str, course: str, model=MODEL):
         self.query = query
         self.course = course
```
Comment (Huy1711) [inline app/llama_sensei/backend/qa/generate_answer.py:26]: initialize embedder here with:
self.embedder = SentenceTransformerEmbeddings(EMBEDDING_LLM, ...)

File: app/llama_sensei/backend/qa/generate_answer.py [hunk 3]
```diff
@@ -1,31 +1,46 @@
+import time
 from datetime import datetime
 
+import numpy as np
 from datasets import Dataset
+from langchain_community.embeddings import SentenceTransformerEmbeddings
+from langchain_community.utilities import DuckDuckGoSearchAPIWrapper
 from langchain_groq import ChatGroq
+from llama_sensei.backend.add_courses.embedding.get_embedding import Embedder
```
Comment (Huy1711) [inline app/llama_sensei/backend/qa/generate_answer.py]: don't import like this, 2 backend will later be separated
use self.embedder instead (see comment below)

File: app/llama_sensei/backend/qa/generate_answer.py [hunk 4]
```diff
@@ -53,44 +68,39 @@ def gen_prompt(self) -> str:
...
 
-        before = datetime.now()
-        res = self.model.invoke(final_prompt)
-        print(f"LLM return time: {datetime.now() - before} seconds")
+        return results
 
-        llm_answer = res['content'] if isinstance(res, dict) else res.content
+    def calculate_context_relevancy(self) -> float:
+        # Embed the query
+        embedder = Embedder()
```
Comment (Huy1711) [inline app/llama_sensei/backend/qa/generate_answer.py]: delete this, use self.embedder instead

File: app/llama_sensei/backend/qa/generate_answer.py [hunk 5]
```diff
@@ -107,44 +117,99 @@ def calculate_faithfulness(self, generated_answer: str) -> float:
         dataset = Dataset.from_dict(data_samples)
 
         # Evaluate faithfulness
-        score = evaluate(dataset, metrics=[faithfulness], llm=model)
+        score = evaluate(
+            dataset, metrics=[faithfulness], llm=model, embeddings=EMBEDDING_LLM
```
Comment (Huy1711) [inline app/llama_sensei/backend/qa/generate_answer.py]: EMBEDDING_LLM -> self.embedder

File: app/llama_sensei/backend/qa/generate_answer.py [hunk 6]
```diff
@@ -107,44 +117,99 @@ def calculate_faithfulness(self, generated_answer: str) -> float:
...
+            self.contexts = [
+                {
+                    "text": result['snippet'],
+                    "metadata": {"link": result['link']},
+                    "embedding": embedder.embed(result['snippet']),
+                }
+                for result in search_results
+            ]
+
+        if indb is True:
```
Comment (Huy1711) [inline app/llama_sensei/backend/qa/generate_answer.py]: please delete "is True"

File: app/llama_sensei/backend/qa/generate_answer.py [hunk 7]
```diff
@@ -107,44 +117,99 @@ def calculate_faithfulness(self, generated_answer: str) -> float:
...
+            self.retrieve_contexts()
+
+        if indb is True or internet is True:
```
Comment (Huy1711) [inline app/llama_sensei/backend/qa/generate_answer.py]: please delete "is True"

**Review by Huy1711 — CHANGES_REQUESTED** [review]

**Review by Huy1711 — APPROVED** [review]
LGTM
